### PR TITLE
Block release v3.0.0

### LINKED ISSUE #680: Lerna bootstrap improvements
With the monorepo transition, we're now able to perform a `lerna bootstrap --ignore-scripts` to link all its packages together, as long as they are in a conformant version range. 

Some dependencies were not symlinked as expected, though, resulting in using to the npm packages (remote). Listed below:
- vm -> block ^2.2.2
- blockchain -> block ~2.2.2
  - block -> tx ^2.1.1

Additionally, I am working on improving how ts-node runs our packages' code, without having to use `npm run build` for each of our dependencies, for each change.

### PR DESCRIPTION
**Pull request by evertonfraga**
This PR initially supersedes https://github.com/ethereumjs/ethereumjs-block/pull/74. Given that this is now a monorepo, I aim to perform the respective integrations in downstream libraries (see diagram below).

<p align="center">

</p>

- [x] Integrate ethereumjs-block v3.0.0 in ethereumjs-blockchain
- [x] Integrate ethereumjs-block v3.0.0 in ethereumjs-vm

Closes #680

### CODE REVIEW

File: packages/block/src/from-rpc.ts [hunk 1]
```diff
@@ -61,6 +60,7 @@ function normalizeTxParams(_txParams: any) {
   // strict byte length checking
   txParams.to = txParams.to ? ethUtil.setLengthLeft(ethUtil.toBuffer(txParams.to), 20) : null
   // v as raw signature value {0,1}
-  txParams.v = txParams.v < 27 ? txParams.v + 27 : txParams.v
+  const v: number = txParams.v
+  txParams.v = v < 27 ? v + 27 : v
```
Comment (holgerd77) [inline packages/block/src/from-rpc.ts:64]: What is happening here?

File: packages/blockchain/src/index.ts [hunk 2]
```diff
@@ -314,7 +312,7 @@ export default class Blockchain implements BlockchainInterface {
    * @hidden
    */
   _setCanonicalGenesisBlock(cb: any): void {
-    const genesisBlock = new Block(null, { common: this._common })
+    const genesisBlock = new Block(undefined, { common: this._common })
```
Comment (holgerd77) [inline packages/blockchain/src/index.ts:315]: Just for understanding: was this a necessary change which triggered some test failure or something before respectively came along with some code changes in `v3.0.0`? Or is this just an adjustment to make things cleaner?

Comment (evertonfraga) [reply packages/blockchain/src/index.ts:315]: This first round of changes is to comply with the linter and `tsc`. There were several compilation failures and Blockchain code was not functional when bumping Block to 3.0.0.

File: packages/blockchain/test/index.ts [hunk 3]
```diff
@@ -585,9 +593,13 @@ test('blockchain test', t => {
...
-        st.equals(head.hash().toString('hex'), genesis.hash().toString('hex'), 'should get head')
-        st.equals(blockchain._heads['head0'].toString('hex'), 'abcd', 'should get state root heads')
-        st.end()
+        if (genesis) {
+          st.equals(head.hash().toString('hex'), genesis.hash().toString('hex'), 'should get head')
+          st.equals(blockchain._heads['head0'].toString('hex'), 'abcd', 'should get state root heads')
+          st.end()
+        } else {
+          st.fail()
+        }
```
Comment (holgerd77) [inline packages/blockchain/test/index.ts:602]: Similar question here on these new conditional block defined yes/no switches: does this just make things cleaner or does it fix some newly triggered errors?

File: packages/blockchain/src/index.ts (hunk 2 above)
Comment (evertonfraga) [reply packages/blockchain/src/index.ts:315]: In this specific case, `null` is not informed as a valid type in the `constructor()`.

```typescript
  constructor(
    data: Buffer | [Buffer[], Buffer[], Buffer[]] | BlockData = {},
    opts: ChainOptions = {},
)
```

File: packages/block/src/from-rpc.ts (hunk 1 above)
Comment (evertonfraga) [reply packages/block/src/from-rpc.ts:64]: TSLint fix. This whole method needs a refactor to remove the magic numbers and is on my to-do list.

File: packages/blockchain/src/index.ts (hunk 2 above)
Comment (evertonfraga) [reply packages/blockchain/src/index.ts:315]: @holgerd77 I am leaning towards allowing `null` in Block constructor. What are your thoughts? 

I can then revert the `null` to `undefined` changes made in this PR.

Comment (holgerd77) [reply packages/blockchain/src/index.ts:315]: Would also have a tendency to keep it "as is".

Comment (evertonfraga) [reply packages/blockchain/src/index.ts:315]: got it. working on it.

File: packages/block/package.json [hunk 4]
```diff
@@ -54,7 +54,6 @@
     "@types/node": "^11.13.4",
     "@types/tape": "^4.2.33",
     "browserify": "^16.2.3",
-    "ethereumjs-blockchain": "^4.0.3",
```
Comment (evertonfraga) [inline packages/block/package.json:57]: removed this circular dependency

File: packages/block/src/block.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import Common from 'ethereumjs-common'
 import * as ethUtil from 'ethereumjs-util'
 import { BN, rlp } from 'ethereumjs-util'
-import { Transaction } from 'ethereumjs-tx'
+import { Transaction, TransactionOptions } from 'ethereumjs-tx'
```
Comment (evertonfraga) [inline packages/block/src/block.ts:4]: We need `TransactionOptions` interface [here](https://github.com/ethereumjs/ethereumjs-vm/pull/682/files#diff-61ac76e0777f844b544ffa824f6ccbdfR82).

File: packages/block/src/block.ts [hunk 6]
```diff
@@ -23,25 +23,30 @@ export class Block {
    * Creates a new block object
    *
    * @param data - The block's data.
-   * @param opts - The network options for this block, and its header, uncle headers and txs.
+   * @param chainOptions - The network options for this block, and its header, uncle headers and txs.
    */
   constructor(
     data: Buffer | [Buffer[], Buffer[], Buffer[]] | BlockData = {},
-    opts: ChainOptions = {},
+    chainOptions: ChainOptions = {},
```
Comment (evertonfraga) [inline packages/block/src/block.ts:30]: Renaming for consistency.

File: packages/block/src/header.ts [hunk 7]
```diff
@@ -283,7 +283,7 @@ export class BlockHeader {
       throw new Error('invalid gas limit')
     }
 
-    if (utils.bufferToInt(parentBlock.header.number) + 1 !== utils.bufferToInt(this.number)) {
+    if (utils.bufferToInt(this.number) - utils.bufferToInt(parentBlock.header.number) !== 1) {
```
Comment (evertonfraga) [inline packages/block/src/header.ts:286]: Solves TSLint operator error.

File: packages/block/test/block.spec.ts [hunk 8]
```diff
@@ -130,109 +144,4 @@ tape('[Block]: block functions', function(t) {
     st.equal(typeof block.toJSON(true), 'object')
     st.end()
   })
-
```
Comment (evertonfraga) [inline packages/block/test/block.spec.ts:133]: removing tests that use package Blockchain.

File: packages/blockchain/src/dbManager.ts [hunk 9]
```diff
@@ -87,11 +87,10 @@ export default class DBManager {
     let body
     try {
       body = await this.getBody(hash, number)
+      return new Block([header, [body], []], { common: this._common })
     } catch (e) {
-      body = [[], []]
+      return new Block([header, [Buffer.from([])], [Buffer.from([])]], { common: this._common })
```
Comment (evertonfraga) [inline packages/blockchain/src/dbManager.ts:92]: changes based on TSLint

File: packages/blockchain/src/index.ts [hunk 10]
```diff
@@ -900,16 +899,19 @@ export default class Blockchain implements BlockchainInterface {
   /**
    * @hidden
    */
-  _delBlock(blockHash: Buffer | typeof Block, cb: any) {
```
Comment (evertonfraga) [inline packages/blockchain/src/index.ts:903]: Now that `Block` is a class, we can drop the `typeof` keyword.
https://github.com/ethereumjs/ethereumjs-vm/pull/682/files#diff-71a1f7e2560231b593e7d6e7805e230fL23

File: packages/blockchain/src/index.ts [hunk 11]
```diff
@@ -900,16 +899,19 @@ export default class Blockchain implements BlockchainInterface {
...
+  _delBlock(blockHash: Buffer | Block, cb: any) {
     const self = this
     const dbOps: any[] = []
     let blockHeader = null
     let blockNumber: any = null
     let parentHash: any = null
     let inCanonical: any = null
+    let hash: Buffer
```
Comment (evertonfraga) [inline packages/blockchain/src/index.ts:909]: refactor: does not reassign `blockHash` argument variable.

File: packages/blockchain/test/index.ts [hunk 12]
```diff
@@ -142,8 +142,12 @@ test('blockchain test', t => {
         st.error(err, 'no error')
         blockchain.getBlock(1, (err?: Error, block?: Block) => {
           st.error(err, 'no error')
-          st.equal(block.hash().toString('hex'), blocks[1].hash().toString('hex'))
-          st.end()
+          if (block) {
```
Comment (evertonfraga) [inline packages/blockchain/test/index.ts:145]: TSLint fix: checking value of `block?` before trying to access it.

File: packages/blockchain/test/index.ts [hunk 13]
```diff
@@ -158,8 +162,12 @@ test('blockchain test', t => {
       st.error(err, 'no error')
       blockchain.getBlock(genesisBlock.hash(), (err?: Error, block?: Block) => {
         st.error(err, 'no error')
-        st.equal(block.hash().toString('hex'), genesisBlock.hash().toString('hex'))
-        st.end()
+        if (block) {
```
Comment (evertonfraga) [inline packages/blockchain/test/index.ts:165]: TSLint fix: checking value of `block?` before trying to access it.

**Comment by evertonfraga** [comment]
The [codecov/patch failure](https://codecov.io/gh/ethereumjs/ethereumjs-vm/compare/b5e295837cc4d5425fdf39baf6730643a840fe6a...d63eb98fdf19063bef4a2f12a4b32b7f9f205554/changes) is expected for this specific case. Note that `codecov/project` is good, meaning that those code paths are at least being executed, anyway.

To remove the circular dependency between `ethereumjs-block` and `ethereumjs-blockchain`, I had to remove the "integration" tests living in `ethereumjs-block`, recently added by @ryanio on https://github.com/ethereumjs/ethereumjs-block/pull/93.

Related commit: https://github.com/ethereumjs/ethereumjs-vm/pull/682/commits/4cba1bac75acdc493e5de8b56f9df77710a6e208.
Related issue: https://github.com/ethereumjs/ethereumjs-vm/issues/746.

File: packages/block/src/from-rpc.ts (hunk 1 above)
Comment (evertonfraga) [reply packages/block/src/from-rpc.ts:69]: @holgerd77 I added a more descriptive comment to explain this magic number.

File: packages/blockchain/src/index.ts (hunk 2 above)
Comment (evertonfraga) [reply packages/blockchain/src/index.ts:315]: Right. In the end, I have changed my mind regarding this.
 
The TypeScript community in general is not much of a fan of `null` values representing `undefined` values. In fact, this is a broad issue on how `null` is much often misused.

So, as the default code from TypeScript version of Block wasn't allowing `null` in their signature, I stick to it, and for JavaScript consumers, though, I added [this check](https://github.com/ethereumjs/ethereumjs-vm/pull/682/files#diff-61ac76e0777f844b544ffa824f6ccbdfR32-R35), which makes sure the method will behave as expected, so if a JS code initializes Block with `data === null`,  `data` will become its [null-object](https://en.wikipedia.org/wiki/Null_object_pattern) equivalent with value `{}`.

The impact in our TS code of using `new Block(undefined, {})` is super low, only used in Block tests.

File: lerna.json [hunk 14]
```diff
@@ -2,10 +2,5 @@
   "packages": [
     "packages/*"
   ],
-  "version": "0.0.0",
-  "command": {
-    "bootstrap": {
-      "ignore": "ethereumjs-block"
-    }
-  }
```
Comment (evertonfraga) [inline lerna.json:10]: ethereumjs-block is now integrated to Lerna workflow, so this change removes it from the ignore list for the bootstrap command.

File: package.json [hunk 15]
```diff
@@ -11,8 +11,9 @@
     }
   },
   "scripts": {
-    "bootstrap": "lerna bootstrap --ignore-scripts && lerna exec npm i --scope=ethereumjs-block",
```
Comment (evertonfraga) [inline package.json:14]: `lerna bootstrap` now installs packages for ethereumjs-block as well, so removing `npm i` here.

File: packages/block/CHANGELOG.md [hunk 16]
```diff
@@ -6,6 +6,116 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+- Update `ethereumjs-common` dependency from `v1.1.0` to `v1.5.0`
+- Update `ethereumjs-tx` dependency from `v1.2.2` to `v2.1.1`
+- Update `ethereumjs-util` dependency from `v5.0.0` to `v6.1.0`
+- Updated test source files to `TypeScript`
+- Signature fix for pre-homestead blocks, see
+  [#67](https://github.com/ethereumjs/ethereumjs-block/issues/67)
+
+[3.0.0]: https://github.com/ethereumjs/ethereumjs-vm/compare/%40ethereumjs%2Fblock%402.2.0...%40ethereumjs%2Fblock%403.0.0
+
+## [2.2.2] - 2019-12-17
```
Comment (evertonfraga) [inline packages/block/CHANGELOG.md:95]: picked 3.0.0, 2.2.1 and 2.2.2 changelog entries from its own branches in ethereumjs-block. Modified a bit 3.0.0 text.

File: packages/block/CHANGELOG.md [hunk 17]
```diff
@@ -24,21 +134,21 @@ PR [#64](https://github.com/ethereumjs/ethereumjs-block/pull/64)
 - Remove `ethereumjs-testing` dependency (much smaller dev dependencies),
   PR [#61](https://github.com/ethereumjs/ethereumjs-block/pull/61)
 
-[2.2.0]: https://github.com/ethereumjs/ethereumjs-vm/compare/%40ethereumjs%2Fvm%402.1.0...%40ethereumjs%2Fvm%402.2.0
+[2.2.0]: https://github.com/ethereumjs/ethereumjs-vm/compare/%40ethereumjs%2Fblock%402.1.0...%40ethereumjs%2Fblock%402.2.0
```
Comment (evertonfraga) [inline packages/block/CHANGELOG.md:137]: fixed comparison link, somehow the `sed` command used during the migration substituted it erroneously.

File: packages/block/CHANGELOG.md [hunk 18]
```diff
@@ -24,21 +134,21 @@ PR [#64](https://github.com/ethereumjs/ethereumjs-block/pull/64)
...
 
 ## [2.1.0] - 2018-10-19
 
 - **Constantinople** support, added difficulty bomb delay (EIP-1234), PR [#54](https://github.com/ethereumjs/ethereumjs-block/pull/54)
 - Updated test data, added Constantinople tests, PR [#56](https://github.com/ethereumjs/ethereumjs-block/pull/56), [#57](https://github.com/ethereumjs/ethereumjs-block/pull/57)
 - Added `timestamp` field to `setGenesisParams()`, PR [#52](https://github.com/ethereumjs/ethereumjs-block/pull/52)
 
-[2.1.0]: https://github.com/ethereumjs/ethereumjs-vm/compare/%40ethereumjs%2Fvm%402.0.1...%40ethereumjs%2Fvm%402.1.0
+[2.1.0]: https://github.com/ethereumjs/ethereumjs-vm/compare/%40ethereumjs%2Fblock%402.0.1...%40ethereumjs%2Fblock%402.1.0
```
Comment (evertonfraga) [inline packages/block/CHANGELOG.md:145]: monorepo migration fix.

File: packages/block/CHANGELOG.md [hunk 19]
```diff
@@ -24,21 +134,21 @@ PR [#64](https://github.com/ethereumjs/ethereumjs-block/pull/64)
...
 
 ## [2.0.1] - 2018-08-08
 
 - Fixes `BlockHeader.prototype.validate()` bug, see PR [#49](https://github.com/ethereumjs/ethereumjs-block/pull/49)
 
-[2.0.1]: https://github.com/ethereumjs/ethereumjs-vm/compare/%40ethereumjs%2Fvm%402.0.0...%40ethereumjs%2Fvm%402.0.1
+[2.0.1]: https://github.com/ethereumjs/ethereumjs-vm/compare/%40ethereumjs%2Fblock%402.0.0...%40ethereumjs%2Fblock%402.0.1
```
Comment (evertonfraga) [inline packages/block/CHANGELOG.md:151]: monorepo migration fix.

File: packages/block/CHANGELOG.md [hunk 20]
```diff
@@ -52,23 +162,23 @@ Changes in detail:
...
+  see [API](https://github.com/ethereumjs/ethereumjs-vm/blob/master/packages/block/docs/index.md) docs
 - Correct block validation for all know hardforks, PR
   [#47](https://github.com/ethereumjs/ethereumjs-block/pull/47), if no hardfork is set validation logic
   is determined by block number in combination with the `chain` set
 - Genesis block initialization depending on the `chain` set (see `ethereumjs-common` for supported chains)
 - Extensive test additions to cover the newly introduced capabilities and changes
 - Fix default value for `nonce` (empty buffer -> `<Buffer 00 00 00 00 00 00 00 00>`), PR [#42](https://github.com/ethereumjs/ethereumjs-block/pull/42)
 
-[2.0.0]: https://github.com/ethereumjs/ethereumjs-vm/compare/%40ethereumjs%2Fvm%401.7.1...%40ethereumjs%2Fvm%402.0.0
+[2.0.0]: https://github.com/ethereumjs/ethereumjs-vm/compare/%40ethereumjs%2Fblock%401.7.1...%40ethereumjs%2Fblock%402.0.0
```
Comment (evertonfraga) [inline packages/block/CHANGELOG.md:173]: monorepo link migration fix.

File: packages/block/CHANGELOG.md [hunk 21]
```diff
@@ -52,23 +162,23 @@ Changes in detail:
 - New initialization parameters `opts.chain` (default: `mainnet`) and `opts.hardfork`
   (default: `null`, block number-based behaviour), PR [#44](https://github.com/ethereumjs/ethereumjs-block/pull/44)
 - Alternatively a `Common` class object can be provided directly with the `opts.common` parameter,
-  see [API](https://github.com/ethereumjs/ethereumjs-block/blob/master/docs/index.md) docs
+  see [API](https://github.com/ethereumjs/ethereumjs-vm/blob/master/packages/block/docs/index.md) docs
```
Comment (evertonfraga) [inline packages/block/CHANGELOG.md:165]: migrating link to monorepo structure.

File: packages/block/package.json [hunk 22]
```diff
@@ -10,7 +10,7 @@
   "scripts": {
     "build": "ethereumjs-config-build",
     "prepublishOnly": "npm run test && npm run build",
-    "coverage": "npx nyc npm run test:node",
```
Comment (evertonfraga) [inline packages/block/package.json:13]: `npx` is irrelevant here as `nyc` is listed as `devDependency`.

File: packages/block/package.json [hunk 23]
```diff
@@ -20,7 +20,7 @@
     "lint": "ethereumjs-config-lint",
     "lint:fix": "ethereumjs-config-lint-fix",
     "test": "npm run test:node && npm run test:browser",
-    "test:node": "npx tape -r ts-node/register test/*.spec.ts",
```
Comment (evertonfraga) [inline packages/block/package.json:23]: `npx` is irrelevant here as `tape` is listed as `devDependency`.

File: packages/block/package.json [hunk 24]
```diff
@@ -70,8 +69,5 @@
     "typedoc-plugin-markdown": "^2.2.17",
     "typescript": "^3.4.3",
     "typestrict": "^1.0.2"
-  },
-  "publishConfig": {
-    "directory": "src"
```
Comment (evertonfraga) [inline packages/block/package.json:75]: We can't do the direct typescript code linking yet without handling the browser tests in a cross-package scope.

File: packages/block/test/block.spec.ts [hunk 25]
```diff
@@ -28,6 +27,21 @@ tape('[Block]: block functions', function(t) {
...
+  })
+
+  t.test('should initialize with null parameters without throwing', function(st) {
+    st.doesNotThrow(function() {
+      const opts = { chain: 'mainnet' }
+      new Block(undefined, opts)
+      st.end()
+    })
+  })
+
```
Comment (evertonfraga) [inline packages/block/test/block.spec.ts:44]: Adding some basic initialization tests

File: packages/blockchain/src/dbManager.ts [hunk 26]
```diff
@@ -83,15 +83,16 @@ export default class DBManager {
...
+    const header: any = (await this.getHeader(hash, number)).raw
     let body
     try {
       body = await this.getBody(hash, number)
     } catch (e) {
       body = [[], []]
     }
 
-    return new Block([header].concat(body), { common: this._common })
+    const blockData = [header].concat(body) as [Buffer[], Buffer[], Buffer[]]
```
Comment (evertonfraga) [inline packages/blockchain/src/dbManager.ts:94]: TS compiler was erroring out on this block of code.

File: packages/block/CHANGELOG.md (hunk 16 above)
Comment (ryanio) [reply packages/block/CHANGELOG.md:95]: looks great 👍

File: packages/block/CHANGELOG.md [hunk 27]
```diff
@@ -6,6 +6,116 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+const Block = require('ethereumjs-block').Block
+```
+
+The library now also comes with a **type declaration file** distributed
+along with the package published.
+
+### Promise-based API
+
+The API of this library is now completely promise-based, the old callback-style
+interface has been dropped.
```
Comment (ryanio) [inline packages/block/CHANGELOG.md:37]: 🔥

**Comment by evertonfraga** [comment]
Thanks for the review, @ryanio!

File: packages/block/src/block.ts [hunk 28]
```diff
@@ -23,25 +23,30 @@ export class Block {
...
   ) {
-    if (opts.common) {
-      if (opts.chain !== undefined || opts.hardfork !== undefined) {
+    // Checking at runtime, to prevent errors down the path for JavaScript consumers.
+    if (data === null) {
+      data = {}
+    }
```
Comment (ryanio) [inline packages/block/src/block.ts:35]: great

File: packages/block/src/from-rpc.ts (hunk 1 above)
Comment (ryanio) [reply packages/block/src/from-rpc.ts:69]: @evertonfraga looks great, would make for a good trivia question 😄

File: packages/blockchain/src/index.ts [hunk 29]
```diff
@@ -14,14 +14,12 @@ import {
...
 } from './util'
 
-const Block = require('ethereumjs-block')
+import { Block, BlockHeader } from 'ethereumjs-block'
 const Ethash = require('ethashjs')
 const Stoplight = require('flow-stoplight')
 const level = require('level-mem')
 const semaphore = require('semaphore')
 
-export type Block = any
```
Comment (ryanio) [inline packages/blockchain/src/index.ts:23]: 🎉

File: packages/blockchain/src/index.ts [hunk 30]
```diff
@@ -470,28 +468,29 @@ export default class Blockchain implements BlockchainInterface {
 
     async.series(
       [
-        verify,
+        async.asyncify(async function() {
```
Comment (ryanio) [inline packages/blockchain/src/index.ts:471]: oh nice 👍

Comment (evertonfraga) [reply packages/blockchain/src/index.ts:471]: a full refactor should be done to simplify all these mixed `async` methods. But that's already planned :)

File: packages/blockchain/test/index.ts (hunk 13 above)
Comment (ryanio) [reply packages/blockchain/test/index.ts:165]: nice yeah sometimes if i'm in a test suite i'll just use `!` to reduce adding bloat. if it is undefined it should error out. the error message is nice though.

File: packages/vm/tests/tester.js [hunk 31]
```diff
@@ -182,6 +182,8 @@ function runTests(name, runnerArgs, cb) {
       const runner = require(`./${name}Runner.js`)
       // Tests for HFs before Istanbul have been moved under `LegacyTests/Constantinople`:
       // https://github.com/ethereum/tests/releases/tag/v7.0.0-beta.1
+
+      // TODO: Replace with Common.lteHardfork('Istanbul')
```
Comment (ryanio) [inline packages/vm/tests/tester.js:186]: 👍 anything blocking this?

Comment (evertonfraga) [reply packages/vm/tests/tester.js:186]: no! but I'd rather make in a separate PR, this one is beyond big. It should be handled with care, testing several test scopes.

**Comment by holgerd77** [comment]
Whew, this is really a biggy! 😁 Congrats on this, feel free to merge in from my side as well.

**Pull request merged**